### ISSUE TITLE
Support declarative configuration

### ISSUE THREAD

**Issue by davibo-oc** [hazelcast/hazelcast-nodejs-client#178]
Currently the only safe way of setting configuration is to do it fully programmatically, i.e.:

```
var Config = require('hazelcast-client').Config;
var config = new Config.ClientConfig()
config.networkConfig.addresses = [{host: '127.0.0.1', port: '5701'}];
var aMapNearCacheConfig = new Config.NearCacheConfig()
aMapNearCacheConfig.inMemoryFormat = Config.InMemoryFormat.BINARY
aMapNearCacheConfig.invalidateOnChange = false
config.nearCacheConfigs = {amap: aMapNearCacheConfig}   
```

Other variations like 

```
config.nearCacheConfigs = {amap: {invalidateOnChange: false}}
```

Or external JSON file like 

```
{   
    "networkConfig": {
        "addresses": [
            {
                "host": "127.0.0.1", 
                "port": "5701"
            }
        ]
    },
    "nearCacheConfigs" : {
        "amap" : {
            "inMemoryFormat" : 1,
            "invalidateOnChange" : true
        }
    }
}
```

will result in losing the default settings in the various items.

Ideally it should be possible to set up configuration via some external, declarative mechanism (e.g. the JSON config file similar to the one above) **with** preserving defaults. Losing any default value is not always a problem, but can lead to unpredictable behaviour.

Additionally - external JSON files don't support enums, you need to use numeric constants to change configuration.

**Comment by gAmUssA** [comment]
@davibo-oc could you elaborate on «will result in losing the default settings in the various items». Data structures are configured on the server. 
Client config has a minor impact on those structures.

**Comment by davibo-oc** [comment]
@gAmUssA 

Consider the following code:

```
var Config = require('hazelcast-client').Config
var hzConfig = require('./hazelcast-client.json')
var defaultConfig = new Config.ClientConfig()
var config = extend(defaultConfig, hzConfig)
console.log(JSON.stringify(config))
```

This takes the default Hazelcast config and extends it with the contents of file `hazelcast-client.json`

Let's assume that I want the following near cache config:

```
{   
    "networkConfig": { /*whatever*/    },
    "nearCacheConfigs" : {
        "amap" : {
            "invalidateOnChange" : true
        }
    }
}```

Now on the print line this is what I'll get:

```
{
  "properties": {
    "hazelcast.client.heartbeat.interval": 5000,
    "hazelcast.client.heartbeat.timeout": 60000
  },
  "networkConfig": { /* whatever */ },
  "customCredentials": null,
  "listeners": {
    "lifecycle": []
  },
  "serializationConfig": { /* whatever */ },
  "reliableTopicConfigs": {
    "default": {
      "readBatchSize": 25,
      "overloadPolicy": 2
    }
  },
  "nearCacheConfigs": {
    "amap": {
      "invalidateOnChange": true
    }
  }
}
```

Compare this with a fully programmatic configuration:

```
{
  "properties": {
    "hazelcast.client.heartbeat.interval": 5000,
    "hazelcast.client.heartbeat.timeout": 60000
  },
  "networkConfig": { /* whatever*/ },
  "customCredentials": null,
  "listeners": {
    "lifecycle": []
  },
  "serializationConfig": { /* whatever */ },
  "reliableTopicConfigs": {
    "default": {
      "readBatchSize": 25,
      "overloadPolicy": 2
    }
  },
  "nearCacheConfigs": {
    "amap": {
      "name": "default",
      "invalidateOnChange": false,
      "maxIdleSeconds": 0,
      "inMemoryFormat": 1,
      "timeToLiveSeconds": 0,
      "evictionPolicy": 0,
      "evictionMaxSize": 9007199254740991,
      "evictionSamplingCount": 8,
      "evictionSamplingPoolSize": 16
    }
  }
}
```

Now the reason we've noticed this is that when `inMemoryFormat` is `undefined` in the near cache config it results in inconsistent behaviour regarding the value returned. There are other problems I've encountered. For example losing `properties` section results in crashing the client

```
LoggingService_1.LoggingService.initialize(this.config.properties['hazelcast.logging']);
                                                                         ^
TypeError: Cannot read property 'hazelcast.logging' of undefined
```

The bottom line is that there are sections of the code that seem to rely on the default values in the configuration, which will be lost if I use JSON fragments to specify the sections I'm interested in changing. 

This is true even if I use inline JSON like:

```
config.nearCacheConfigs = {amap: {invalidateOnChange: false}}
```

Possible solutions
1 Remove reliance of default config values in the code **OR**
1 Support a fully recursive merge of JSON-style config and the defaults

**Comment by mustafaiman** [comment]
fixed by https://github.com/hazelcast/hazelcast-nodejs-client/pull/259
